ramips: add support for DLINK DWR-116-A1 and A2

CHKDSK88 opened the pull request:
The DWR-116-A1 and A2 Wireless Routers are based on the MT7620N SoC.
    
    Specification:
    
    -MediaTek MT7620N (580 Mhz)
    -32 MB of RAM
    -8 MB of FLASH
    -802.11bgn radio
    -5x 10/100 Mbps Ethernet (1 WAN and 4 LAN)
    -2x external, non-detachable antennas
    -UART (J1) header on PCB (57600 8n1)
    -6x LED (GPIO-controlled), 2x button
    -JBOOT bootloader
    
    Known issues:
    WAN LED is drived by uartl tx pin. I decide to use this pin as uartl tx pin.

    Known strange things:
    D-link make config partition diffrent in all verions. In A2PL and A2EU difference is only 1 byte. Offset in 
    dtb is result different calibration data in A1, A2EU and A2PL.
    
    Installation:
    Apply factory image via d-link http web-gui.

    How to revert to OEM firmware:
    1.) Push the reset button and turn on the power. Wait until LED start blinking (~10sec.)
    2.) Upload original factory image via JBOOT http (IP: 192.168.123.254)
    3.) If http doesn't work, it can be done with curl command:
        curl -F FN=@XXXXX.bin http://192.168.123.254/upg
        where XXXXX.bin is name of firmware file.
 
    
    Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
ramips: add support for D-Link DWR-116-A1/2

The DWR-116-A1/2 Wireless Router is based on the MT7620N SoC.

Specification:

-MediaTek MT7620N (580 Mhz)
-32 MB of RAM
-8 MB of FLASH
-802.11bgn radio
-5x 10/100 Mbps Ethernet (1 WAN and 4 LAN)
-2x external, non-detachable antennas
-UART (J1 in A1, JP1 in A2) header on PCB (57600 8n1)
-6x LED (GPIO-controlled), 2x button
-JBOOT bootloader

Known issues:
WAN LED is drived by uartl tx pin. I decide to use this pin as uartl tx pin.

Installation:
Apply factory image via http web-gui.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

mkresin commented on target/linux/ramips/dts/DWR-116-A1.dts:
```diff
@@ -0,0 +1,139 @@
+			gpios = <&gpio2 3 GPIO_ACTIVE_LOW>;
+		};
+
+		/*wan {  /*WAN LED IS UART TX*/
```
Drop this one and add a note to the commit message, that the wan led is using the uartlite pins and you prefer to have a serial console rather than a WAN LED.

xabolcs replied:
@mkresin , are there any (easy) way to reconfigure the pins at runtime?

mkresin replied:
No, it isn'T possible to reconfigure pins during runtime.

mkresin commented on target/linux/ramips/dts/DWR-116-A1.dts:
```diff
@@ -0,0 +1,139 @@
+
+&ethernet {
+	mtd-mac-address = <&config 0xe291>;
+	mediatek,portmap = "llllw";
```
if you add here:

```
	pinctrl-names = "default";
	pinctrl-0 = <&ephy_pins>;
```

you can let the hardware control the leds. It would allow you to drop the ucidef_set_led_switch from userspace. You need to drop the `ephy` from the gpio mux as well. Since it is a rather controversial discussion if LEDs should be controllable from userspace or handled by hardware, I only mention how it can be done. Handling LEDs by hardware would mean less cpu cycles wastes for userspace LED handling.

CHKDSK88 replied:
@mkresin 
I agree with You.

All changes are done.

mkresin commented:
Your hardware description in the commit message is insufficient. Please have a look at [recent additions](https://git.lede-project.org/?p=source.git&a=search&h=HEAD&st=commit&s=add+support+for) for some better examples. Without more knowledge about the hardware it is impossible to do a proper review.

Be warned that I don't like the fact that the bootloader need to be replaced. We have a least one more board which is pending because of the stock image format (https://github.com/lede-project/source/pull/1279).

I don't like it for the following reasons:

- doing it wrong would brick the board, without an easy way of recovery
- the bootloader is provided from a third party source and might vanish at some time, which makes support for the DWR-116-A1 somehow dead code
- the bootloader is available as binary only
- there are plenty of mt7620 boards support, I don't really see the need to add another one which required such a risky installation

mkresin reviewed (changes requested):
Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

CHKDSK88 commented:
@mkresin 
I agree, that much better is using original bootloader.
I use the same method, whitch is used in: [DWR-512-B](https://wiki.openwrt.org/toh/d-link/d-link_dwr-512_b). Custom u-boot is made from sources based on mediatek sdk. I made only configuration for this board. The major changes can be quoted in wiki to preserve from vanish.

CHKDSK88 committed:
firmware-utils: mkdlinkfw: new tool for D-Link devices with JBOOT bootloader

Tested on D-Link DWR-116.
Based on mktplinkfw.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

xabolcs commented:
@mkresin, how about making two flavours? One for the `U-Boot` bootloader (for the brave guys and girls) and one for the OEM `JBOOT` bootloader?

@CHKDSK88, it would be nice if the JBOOT flavour would have prepared for [extroot](https://wiki.openwrt.org/doc/howto/extroot#automatic_setup).

CHKDSK88 commented:
@xabolcs, I'am not sure if U-Boot version is required after JBOOT breaking. It's risky and IMO pointless in present situation.
I plan to change this PR for JBOOT verion only. But I can make both versions.

CHKDSK88 commented:
I make some cleanup and now is supported jboot bootloader. But it need my two additional PR.

xabolcs commented:
@CHKDSK88, you are right! The JBOOT version has even more free space! 👍

xabolcs commented:
I have a DWR-116 with hardware revision `A2`.

It's PCB layout is slightly different from `A1`'s, but they share the same OEM firmware, and the factory image built from this PR (with those referenced commits) flashes and boots flawlessly.

It have the same SoC and flash & RAM size. 
Should it be distinguished? How should be noted that this PR will support it?

@CHKDSK88, almost offtopic, but starting the wifi (with `psk2` encryption) traces (and doesn't work) for me on this `A2` hardware:
```
[  331.827052] ------------[ cut here ]------------
[  331.836307] WARNING: CPU: 0 PID: 1211 at backports-2017-11-01/drivers/net/wireless/ralink/rt2x00/rt2800lib.c:5233 rt2800_rfkill_poll+0x15b8/0x1908 [rt2800lib]
[  331.864560] Not supported RF chipet 2 for VCO recalibration
[  331.875315] Modules linked in:
[  331.881386]  rt2800soc rt2800pci rt2800mmio rt2800lib pppoe ppp_async rt2x00soc rt2x00pci rt2x00mmio rt2x00lib pppox ppp_generic mt76x2e mt7603e mt76 mac80211 iptable_nat ipt_REJECT ipt_MASQUERADE cfg80211 xt_time xt_tcpudp xt_state xt_nat xt_multiport xt_mark xt_mac xt_limit xt_conntrack xt_comment xt_TCPMSS xt_REDIRECT xt_LOG slhc nf_reject_ipv4 nf_nat_redirect nf_nat_masquerade_ipv4 nf_conntrack_ipv4 nf_nat_ipv4 nf_nat nf_log_ipv4 nf_defrag_ipv4 nf_conntrack_rtcache nf_conntrack iptable_mangle iptable_filter ip_tables crc_itu_t crc_ccitt compat ip6t_REJECT nf_reject_ipv6 nf_log_ipv6 nf_log_common ip6table_mangle ip6table_filter ip6_tables x_tables eeprom_93cx6 leds_gpio ehci_platform ehci_hcd gpio_button_hotplug usbcore nls_base usb_common
[  332.010263] CPU: 0 PID: 1211 Comm: hostapd Not tainted 4.9.77 #0
[  332.022232] Stack : 8045751a 00000034 00000000 00000001 81832c04 803ff267 803aaf1c 000004bb
[  332.038914]         80453660 00001471 0000003f 000008ff 00000021 8004f0d0 803b1490 80400000
[  332.055597]         00000003 00001471 803af130 80e51ad4 00000021 800878bc 8045751a 00000092
[  332.072267]         00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[  332.088949]         00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[  332.105628]         ...
[  332.110490] Call Trace:
[  332.115384] [<8000e478>] show_stack+0x54/0x88
[  332.124073] [<800248a0>] __warn+0xe4/0x118
[  332.132246] [<80024904>] warn_slowpath_fmt+0x30/0x3c
[  332.142159] [<80dcdc34>] rt2800_rfkill_poll+0x15b8/0x1908 [rt2800lib]
[  332.155008] ---[ end trace bb51ae17337c1d03 ]---
```

Used wireless config:
```root@OpenWrt:~# cat /etc/config/wireless

config wifi-device 'radio0'
        option type 'mac80211'
        option channel '11'
        option hwmode '11g'
        option path 'platform/10180000.wmac'
        option htmode 'HT20'
        option disabled '0'

config wifi-iface 'default_radio0'
        option device 'radio0'
        option network 'lan'
        option mode 'ap'
        option ssid 'OpenWrt'
        option encryption 'psk2'
        option key 'OpenWrtOpenWrt'

```
Steps to reproduce:
1. create simple wireless config like above
1. `wifi`

CHKDSK88 commented:
@xabolcs 

I think, it might be little diffrent config memory offset map. I had those problems durring generic image tests.
Could You make for me some things?
1. Check if MAC adress is random.
2. Make hexdump -C /dev/mtd5 and send to result to me?

xabolcs commented:
@CHKDSK88 

1. yeah, it is random
2. sure, will send it privately

xabolcs commented:
The `eth0` changes it's MAC address every time after `firstboot`:
```
[    0.715465] mtk_soc_eth 10100000.ethernet eth0 (uninitialized): port 0 link up (100Mbps/Full duplex)
[    0.733796] mtk_soc_eth 10100000.ethernet: generated random MAC address 1e:b9:8b:d0:a8:c1
```

CHKDSK88 commented:
Xabolcs have a A2 EU version, it have diffrent offset in config memory.

CHKDSK88 commented:
Added status led support. Thanks @xabolcs for help with this.

CHKDSK88 commented:
PR is now kernel 4.14 ready.

fid0did0 commented:
Hi,
I successfully tested the commit b416906 on the DWR-921 (embedding the mt7620n as well).
With this patch the wan port is properly enabled and work as expected.

My advice is to decouple this commit from this PR because is not related only to the DWR-116 but to all the devices embedding the mt7620n.

In my opinion is better to address this commit in a new dedicated PR.

Bye.

CHKDSK88 committed:
firmware: add JBOOT based devices config extractor

Adds tool to extract MAC and pre-calibration data required for JBOOT
based D-Link routers.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 commented:
@mkresin 

Should I split this PR?

mkresin commented:
> Should I split this PR?

No keep it as it is. I will cherry pick stuff.

In fact I was about commiting parts of the PR but there are still issues. Please export your Jboot patch and fix the errors reported by `./scripts/checkpatch.pl path/to/patch.patch`

Please fix the compiler warnings for mkdlinkfw as well:

```
$ make tools/firmware-utils/{clean,compile} V=s

gcc -O2 -I/openwrt/dev/staging_dir/host/include  -include endian.h -L/openwrt/dev/staging_dir/host/lib  -o /openwrt/dev/build_dir/host/firmware-utils/bin/mkdhpimg src/mkdhpimg.c src/buffalo-lib.c  -Wall
src/mkdhpimg.c: In function 'main':
src/mkdhpimg.c:68:2: warning: ignoring return value of 'read', declared with attribute warn_unused_result [-Wunused-result]
  read(in, &buf[DHP_HEADER_SIZE], in_st.st_size);
  ^
src/mkdhpimg.c:79:2: warning: ignoring return value of 'write', declared with attribute warn_unused_result [-Wunused-result]
  write(out, buf, size);
  ^
gcc -O2 -I/openwrt/dev/staging_dir/host/include  -include endian.h -L/openwrt/dev/staging_dir/host/lib  -o /openwrt/dev/build_dir/host/firmware-utils/bin/mkdlinkfw src/mkdlinkfw.c src/mkdlinkfw-lib.c  -lz -Wall
src/mkdlinkfw.c: In function 'find_auh_headers':
src/mkdlinkfw.c:242:12: warning: format '%X' expects argument of type 'unsigned int', but argument 2 has type 'long int' [-Wformat=]
     printf("Find proper AUH header at: 0x%X!\n",
            ^
src/mkdlinkfw.c:232:6: warning: unused variable 'buf_size' [-Wunused-variable]
  int buf_size = inspect_info.file_size;
      ^
src/mkdlinkfw.c: In function 'check_stag_header':
src/mkdlinkfw.c:295:10: warning: format '%X' expects argument of type 'unsigned int', but argument 2 has type 'long int' [-Wformat=]
   printf("Find proper STAG header at: 0x%X!\n",
          ^
src/mkdlinkfw.c: In function 'check_sch2_header':
src/mkdlinkfw.c:322:28: warning: passing argument 2 of 'crc32' from incompatible pointer type [-Wincompatible-pointer-types]
  if (crc32_tmp == crc32(0, header, header->header_length)) {
                            ^
In file included from src/mkdlinkfw.c:28:0:
/usr/include/zlib.h:1600:23: note: expected 'const Bytef * {aka const unsigned char *}' but argument is of type 'struct sch2_header *'
 ZEXTERN uLong ZEXPORT crc32   OF((uLong crc, const Bytef *buf, uInt len));
                       ^
src/mkdlinkfw.c:324:10: warning: format '%X' expects argument of type 'unsigned int', but argument 2 has type 'long int' [-Wformat=]
   printf("Find proper SCH2 header at: 0x%X!\n",
          ^
src/mkdlinkfw.c:328:16: warning: pointer targets in passing argument 2 of 'crc32' differ in signedness [-Wpointer-sign]
       crc32(0, (char *)header + header->header_length,
                ^
In file included from src/mkdlinkfw.c:28:0:
/usr/include/zlib.h:1600:23: note: expected 'const Bytef * {aka const unsigned char *}' but argument is of type 'char *'
 ZEXTERN uLong ZEXPORT crc32   OF((uLong crc, const Bytef *buf, uInt len));
                       ^
src/mkdlinkfw.c: In function 'fill_sch2':
src/mkdlinkfw.c:409:33: warning: pointer targets in passing argument 2 of 'crc32' differ in signedness [-Wpointer-sign]
  header->image_crc32 = crc32(0, kernel_ptr, kernel_info.file_size);
                                 ^
In file included from src/mkdlinkfw.c:28:0:
/usr/include/zlib.h:1600:23: note: expected 'const Bytef * {aka const unsigned char *}' but argument is of type 'char *'
 ZEXTERN uLong ZEXPORT crc32   OF((uLong crc, const Bytef *buf, uInt len));
                       ^
src/mkdlinkfw.c:414:34: warning: pointer targets in passing argument 2 of 'crc32' differ in signedness [-Wpointer-sign]
  header->rootfs_crc32 = crc32(0, rootfs_ptr, rootfs_info.file_size);
                                  ^
In file included from src/mkdlinkfw.c:28:0:
/usr/include/zlib.h:1600:23: note: expected 'const Bytef * {aka const unsigned char *}' but argument is of type 'char *'
 ZEXTERN uLong ZEXPORT crc32   OF((uLong crc, const Bytef *buf, uInt len));
                       ^
src/mkdlinkfw.c:419:34: warning: passing argument 2 of 'crc32' from incompatible pointer type [-Wincompatible-pointer-types]
  header->header_crc32 = crc32(0, header, header->header_length);
                                  ^
In file included from src/mkdlinkfw.c:28:0:
/usr/include/zlib.h:1600:23: note: expected 'const Bytef * {aka const unsigned char *}' but argument is of type 'struct sch2_header *'
 ZEXTERN uLong ZEXPORT crc32   OF((uLong crc, const Bytef *buf, uInt len));
                       ^
src/mkdlinkfw.c: In function 'build_fw':
src/mkdlinkfw.c:547:3: warning: 'auh_header_kernel' may be used uninitialized in this function [-Wmaybe-uninitialized]
   fill_auh(auh_header_kernel, writelen - AUH_SIZE);
   ^
src/mkdlinkfw-lib.c: In function 'read_to_buf':
src/mkdlinkfw-lib.c:95:2: warning: ignoring return value of 'fread', declared with attribute warn_unused_result [-Wunused-result]
  fread(buf, fdata->file_size, 1, f);
  ^
```

CHKDSK88 commented:
@mkresin 

Done.

mkresin commented:
Something is strange with the different mtd offsets between the three boards supported with this PR:

```
A1:   eeprom config@0xe291, mac address: config@0xe084
A2PL: eeprom config@0xe2a1, mac address: config@0xe094
A2EU: eeprom config@0xe2a2, mac address: config@0xe095
```

Never seen something like that before and it rather looks like there arn't fixed offsets for eeprom and mac address.

Could it be that the config partition is some kind of filesystem? Or something similar to the u-boot environment, which is a list of NULL terminated strings without a fixed order.

CHKDSK88 commented:
@mkresin 
Yes, this is strange.  I can't figure out how D-link make this.

Links to full stock dumps:
https://www.dropbox.com/s/n51unpssyh0rtvi/d-link_dwr-116-a2pl-stock.bin
https://www.dropbox.com/s/56bdvythxikfw9o/d-link_dwr-116-a1-stock.bin
https://www.dropbox.com/s/s7rjclp80mn7tvs/d-link_dwr-116-a2eu-stock.bin

xabolcs commented:
The only software difference is those addresses between those hardware revisions. You will see, the eeprom and mac addresses share the same pattern.

Are there any way to detect those patterns and find the offsets programmatically to unify DWR-116 support?

xabolcs commented:
@CHKDSK88, could you paste here the `info` output from JBOOT recovery console from those `A1` and `A2PL` revisions posted above?

Just for the patterns! 

Here is it, for the `A2EU` dump above:

```
$ telnet 192.168.123.254
Trying 192.168.123.254...
Connected to 192.168.123.254.
Escape character is '^]'.
info
# info
MODEL_NAME: DWR-116
ROMID: DLK6E3803001
ECN_NO: 201510200003E
FMID: 6E38
PV: 1
IP: 192.168.123.254
BEID: 6C:72:20:A5:58:35
PSEC: BC7FE000 80654500
```

mkresin commented:
I had a brief look at the dumps. First of all your EEPROM offsets are correct, but can't be used as it is now.

The EEPROM for the mt7620 build in wifi starts with the magic `20 76 05 01`. The magic can be found at the following offsets of the config partitions:

A1: `0xe084`
A2EU: `0xe095`
A2PL: `0xe094`

At offset `0x28` relative to the beginning of the EEPROM, the wireless mac address is stored. We are usually using this mac address as base for all other mac addresses used in the system.

But in your case the mac addresses in the EEPROMs aren't correctly populated and a default ralink mac address is used. The default ralink mac address is the same for all boards, which will cause nasty issues if >1 board is used in the same network.

It indeed looks like there is a dynamic structure in the config partition. Given your `info` output, the structure starts at offset `0xe000` for all boards and has a dynamic size. The systems mac address is part that structure. Most likely the EEPROM is part of that structure as well. If not, it at least depends on the size of the structure. 

What is missing, is a parser for that structure similar to the `info` command from the dlink firmware. We can add and use such a userspace program, as we have done it with the [AVM tffs tools](https://github.com/openwrt/openwrt/tree/a367645f23d2ed93ea29c7237fa1b2d6c3ded7e4/package/utils/fritz-tools) (grep for `fritz_tffs` and `fritz_cal_extract` in the OpenWrt tree).

To reliable extract the EEPROM, you need to go the userland way (target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom), using a dlink/jboot config partition parser tool . The ralink,mtd-eeprom/mtd-mac-address devicetree property can be only used with fixed and never changing offsets. You can patch a correct wireless mac via the same file. 

The ethernet interface mac addresses can be set via 02_network.

I wouldn't be surprised if the DWR-921, @fid0did0 is working on, uses the same structure.

CHKDSK88 commented:
@mkresin 

I prepare the config extractor. The offset after "DLK" is constant (i look also in DWR-922 stock image), so I make the program in user space.

But is one problem: On stock firmware WLAN interface have the same MAC as LAN. LAN have WAN MAC +1. I see four possible solutions:
1. Set WLAN MAC as WAN MAC +2.
2. Set WLAN as random once and keep it.
3. On every start set random WLAN MAC.
4. Set WLAN MAC the same as LAN MAC (WAN +1).

Which solution is the most Open-WRT way?

mkresin commented:
> On stock firmware WLAN interface have the same MAC as LAN

it is fine to do the same.

> The offset after "DLK" is constant (i look also in DWR-922 stock image), so I make the program in user space.

It isn't exactly what I meant with a parser. I mean searching for a well known string and assuming a fixed offset between the string and the data you are looking for sounds really error prone.

It looks like the single values/entries are separated by a pattern:

```
02 02 24 2B 4D 1B 49 08 9C 02 00 00 FF FF A7 AC 43 53 1F 8E 8C 02 00 00 00 00 00 00 8C 02 00 00 18
42 00 80 0E 00 32 30 31 35 31 30 32 30 30 30 30 33 45 00 19
42 00 80 01 00 01 20
42 00 80 08 00 44 57 52 2D 31 31 36 00 26
42 00 80 03 00 41 32 00 27
42 00 80 03 00 41 32 00 21
42 00 80 02 00 02 05 22
42 00 80 04 00 C0 A8 7B FE 10
42 00 80 0D 00 44 4C 4B 36 45 33 38 30 33 30 30 31 00 01
42 00 80 02 00 38 6E 02
42 00 80 01 00 07 00
F5 00 80 00 02 20 76
```

The forth value could be a size argument:

```
42 00 80 0E 00   32 30 31 35 31 30 32 30 30 30 30 33 45 00 /* 201510200003E (13 char + 0x00 == 0x0E) */
42 00 80 0D 00   44 4C 4B 36 45 33 38 30 33 30 30 31 00 /* DLK6E3803001 (12 chars +0x00 == 0x0D ) */
42 00 80 08 00   44 57 52 2D 31 31 36 00 /* DWR-116 (7 chars + 0x00 = 0x08) */
```

Instead of reverse engineering the format, I hope there is some code in the GPL dump that explains the format.

CHKDSK88 commented:
Thank to You, I found the pattern:

```
02 02 24 2B -> Magic
0D D6 B2 07 -> jboot timestamp 
8B 02 00 00 -> data size (Start at config + 0x10)
FF FF -> ???
8E F2 -> ~jboot checksum
43 53 CE 83 -> ???
7B 02 00 00 -> data size (Start at config + 0x20)
00 00 00 00 -> ???
7B 02 00 00 -> data size (Start at config + 0x20)
18 -> some magic? (first footer?)
42 00 80 0D 00 32 30 31 34 30 37 32 31 39 39 39 39 00 19 -> HEADER (42 -> config data (?), 000D -> size): 42 00 80 0D 00 DATA: 32 30 31 34 30 37 32 31 39 39 39 39 FOOTER: 19
42 00 80 01 00 01 20 -> HEADER (42 -> config data (?), 0001 -> size): 42 00 80 01 00 DATA: 01 FOOTER: 20
42 00 80 0A 00 44 57 52 2D 31 31 36 41 31 00 21 -> HEADER (42 -> config data (?), 000A -> size): 42 00 80 0A 00 DATA: 44 57 52 2D 31 31 36 41 31 00 FOOTER: 20
42 00 80 02 00 02 05 22 -> the same pattern
42 00 80 04 00 C0 A8 7B FE 10 -> the same pattern
42 00 80 0D 00 44 4C 4B 36 45 33 38 30 33 30 30 31 00 01 -> the same pattern
42 00 80 02 00 38 6E 02 -> the same pattern
42 00 80 01 00 07 00 -> the same pattern
F5 00 80 00 02 20 76 05 01 00 0C 43 76 20 20 FF FF  -> HEADER (F5 -> wireless eeprom data (?), 0200 -> size): F5 00 80 00 02 DATA: .........
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF 00 0C 43 76 
20 77 00 0C 43 76 20 66 22 0C 00 00 FF 01 6E 01 
55 77 A8 AA 8C 88 FF FF 0A 00 00 00 00 00 00 00 
00 00 FF FF FF FF 0D 0D 0D 0D 0C 0C 0C 0C 0C 0B 
0B 0B 0B 0B 15 14 13 12 10 0F 0E 10 12 13 15 15 
15 15 80 FF FF FF 80 FF FF FF 00 00 FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF 20 FF FF FF FF FF FF FF FF FF FF FF 
FF FF 08 08 08 08 04 00 07 07 04 00 08 08 04 00 
07 07 04 00 FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF 62 -> FOOTER: 62
42 00 80 01 00 00 30 -> the same pattern
42 00 80 06 00 EC 22 80 C4 40 C0 00 -> the same pattern (MAC, last phrase)
00 00 00 -> align to  four 
```

I check four different stock images. MAC is allways last phrase, before alignment. EEPROM wireless data start from F5. I try to find more info in D-link GPL code.

xabolcs commented:
Just found in the GPL roofs, in the `/usr/etc/rcS` file a related line. Note the `0x7FE000` address:
```
getpsec -f 0x7FE000 /tmp/csman/pre1.dat
 ```

It creates a dump ( which is for that `A2EU` above):
```$ hexdump -C pre1.dat
00000000  43 53 1f 8e 8c 02 00 00  00 00 00 00 8c 02 00 00  |CS..............|
00000010  18 42 00 80 0e 00 32 30  31 35 31 30 32 30 30 30  |.B....2015102000|
00000020  30 33 45 00 19 42 00 80  01 00 01 20 42 00 80 08  |03E..B..... B...|
00000030  00 44 57 52 2d 31 31 36  00 26 42 00 80 03 00 41  |.DWR-116.&B....A|
00000040  32 00 27 42 00 80 03 00  41 32 00 21 42 00 80 02  |2.'B....A2.!B...|
00000050  00 02 05 22 42 00 80 04  00 c0 a8 7b fe 10 42 00  |..."B......{..B.|
00000060  80 0d 00 44 4c 4b 36 45  33 38 30 33 30 30 31 00  |...DLK6E3803001.|
00000070  01 42 00 80 02 00 38 6e  02 42 00 80 01 00 07 00  |.B....8n.B......|
00000080  f5 00 80 00 02 20 76 05  01 00 0c 43 76 20 c0 ff  |..... v....Cv ..|
00000090  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
000000a0  ff ff ff ff ff ff ff ff  ff ff ff ff ff 00 0c 43  |...............C|
000000b0  76 20 77 00 0c 43 76 20  66 22 0c 00 00 00 01 5b  |v w..Cv f".....[|
000000c0  01 55 77 a8 aa 8c 88 ff  ff 0a 00 00 00 00 00 00  |.Uw.............|
000000d0  00 00 00 ff ff ff ff 0c  0c 0b 0b 0a 0a 09 0a 0a  |................|
000000e0  0b 0b 0c 0c 0c 11 10 0f  0e 0c 0b 0a 0b 0c 0d 0e  |................|
000000f0  0f 10 10 80 ff ff ff 80  ff ff ff 00 00 ff ff ff  |................|
00000100  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000150  ff ff ff ff ff 20 ff ff  ff ff ff ff ff ff ff ff  |..... ..........|
00000160  ff ff ff 08 08 08 08 04  00 07 07 04 00 08 08 04  |................|
00000170  00 07 07 04 00 ff ff ff  ff ff ff ff ff ff ff ff  |................|
00000180  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000280  ff ff ff ff ff 62 42 00  80 01 00 00 30 42 00 80  |.....bB.....0B..|
00000290  06 00 6c 72 20 a5 58 35  00 00 00 00              |..lr .X5....|
0000029c
```

You could see it has offset `0x10`.

There is a `src/include/psec.h` in the GPL dump, which looks promising to me (but I'm not a C/C++ programmer).

CHKDSK88 commented:
Possibile solution:

```
02 02 24 2B -> Magic
0D D6 B2 07 -> jboot timestamp 
8B 02 00 00 -> data size (Start at config + 0x10)
FF FF -> image jboot checksum (?)
8E F2 -> ~jboot checksum
43 53 -> Magic 
CE 83 -> Image jboot checksum (?)
7B 02 00 00 -> data size (Start at config + 0x20) (body_lenght)
00 00 00 00 -> (body_encoding)
7B 02 00 00 -> data size (Start at config + 0x20) (raw_length)
18 42 00 80 0D 00 32 30 31 34 30 37 32 31 39 39 39 39 00 -> HEADER (4218-> config data (id 18 - build time (?) ), 000D -> size): 18 42 00 80 0D 00 DATA: 32 30 31 34 30 37 32 31 39 39 39 39
19 42 00 80 01 00 01 -> HEADER (4219 -> config data (id 19), 0001 -> size): 19 42 00 80 01 00 DATA: 01
20 42 00 80 0A 00 44 57 52 2D 31 31 36 41 31 00 -> HEADER (4220 -> config data (id 20 - model), 000A -> size): 20 42 00 80 0A 00 DATA: 44 57 52 2D 31 31 36 41 31 00
21 42 00 80 02 00 02 05 -> the same pattern 
22 42 00 80 04 00 C0 A8 7B FE -> the same pattern
10 42 00 80 0D 00 44 4C 4B 36 45 33 38 30 33 30 30 31 00 -> the same pattern
01 42 00 80 02 00 38 6E -> the same pattern
02 42 00 80 01 00 07 -> the same pattern
00 F5 00 80 00 02 20 76 05 01 00 0C 43 76 20 20 FF FF  -> HEADER (F500 -> wireless eeprom data (id 0), 0200 -> size): 00 F5 00 80 00 02 DATA: .........
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF 00 0C 43 76 
20 77 00 0C 43 76 20 66 22 0C 00 00 FF 01 6E 01 
55 77 A8 AA 8C 88 FF FF 0A 00 00 00 00 00 00 00 
00 00 FF FF FF FF 0D 0D 0D 0D 0C 0C 0C 0C 0C 0B 
0B 0B 0B 0B 15 14 13 12 10 0F 0E 10 12 13 15 15 
15 15 80 FF FF FF 80 FF FF FF 00 00 FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF 20 FF FF FF FF FF FF FF FF FF FF FF 
FF FF 08 08 08 08 04 00 07 07 04 00 08 08 04 00 
07 07 04 00 FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
FF FF FF FF
62 42 00 80 01 00 00 30 -> the same pattern
30 42 00 80 06 00 EC 22 80 C4 40 C0 -> the same pattern (MAC, id 30)
00 00 00 00 -> end
```

It fits to 116 A1, A2EU PL and 922. 
@fid0did0 Could You check in DWR-921 config partition?

I will make an data extractor.

xabolcs commented:
> `62 42 00 80 01 00 00 30 -> the same pattern`

There is a typo in your Section id 62: there is an extra `30` (from the following section). ;)

fid0did0 commented:
Similar structure on DWR 921 with some differences:

```
0000e000: 02 02 24 2b 03 2b 0d 08 e3 02 00 00 ff 00 e7 9b  ..$+.+..........
0000e010: 43 53 7c 90 d3 02 00 00 00 00 00 00 d3 02 00 00  CS|.............
0000e020: 18 42 00 80 0e 00 32 30 31 35 30 34 32 31 30 30  .B....2015042100
0000e030: 30 32 45 00 19 42 00 80 01 00 01 20 42 00 80 08  02E..B..... B...
0000e040: 00 44 57 52 2d 39 32 31 00 26 42 00 80 03 00 43  .DWR-921.&B....C
0000e050: 31 00 27 42 00 80 03 00 43 31 00 21 42 00 80 02  1.'B....C1.!B...
0000e060: 00 02 1a 22 42 00 80 04 00 c0 a8 7b fe 10 42 00  ..."B......{..B.
0000e070: 80 0d 00 44 4c 4b 36 45 32 34 31 34 30 30 31 00  ...DLK6E2414001.
0000e080: 01 42 00 80 02 00 24 6e 00 f5 00 80 00 02 20 76  .B....$n...... v
0000e090: 05 01 00 0c 43 76 20 e0 ff ff ff ff ff ff ff ff  ....Cv .........
0000e0a0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e0b0: ff ff ff ff ff ff 00 0c 43 76 20 77 00 0c 43 76  ........Cv w..Cv
0000e0c0: 20 66 22 0c 00 00 00 01 61 01 55 77 a8 aa 8c 88   f".....a.Uw....
0000e0d0: ff ff 0a 00 00 00 00 00 00 00 00 00 ff ff ff ff  ................
0000e0e0: 06 06 07 07 08 08 09 09 09 08 08 08 07 07 07 07  ................
0000e0f0: 07 07 07 06 06 06 06 07 07 07 08 08 80 ff ff ff  ................
0000e100: 80 ff ff ff 00 00 ff ff ff ff ff ff ff ff ff ff  ................
0000e110: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e120: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e130: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e140: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e150: ff ff ff ff ff ff ff ff ff ff ff ff ff ff 20 ff  .............. .
0000e160: ff ff ff ff ff ff ff ff ff ff ff ff 08 08 08 08  ................
0000e170: 04 00 07 07 04 00 08 08 04 00 07 07 04 00 ff ff  ................
0000e180: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e190: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1a0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1b0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1c0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1d0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1e0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1f0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e200: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e210: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e220: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e230: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e240: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e250: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e260: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e270: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e280: ff ff ff ff ff ff ff ff ff ff ff ff ff ff 37 42  ..............7B
0000e290: 00 80 12 00 47 31 53 4a 46 35 30 30 30 34 42 33  ....G1SJF50004B3
0000e2a0: 4a 30 31 00 55 53 30 42 00 80 06 00 e4 6f 13 26  J01.US0B.....o.&
0000e2b0: d5 51 81 42 00 80 11 00 50 55 72 46 35 55 58 52  .Q.B....PUrF5UXR
0000e2c0: 75 52 36 36 50 54 63 6b 00 83 42 00 80 11 00 76  uR66PTck..B....v
0000e2d0: 71 47 65 46 36 45 33 38 32 39 6e 77 53 38 34 00  qGeF6E3829nwS84.
0000e2e0: 82 42 00 80 09 00 36 32 35 36 37 39 32 38 00 00  .B....62567928..
0000e2f0: 00 00 00 ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
```
and here the partially decoded:
```
02 02 24 2b 
03 2b 0d 08 
e3 02 00 00 
ff 00 
e7 9b
43 53 
7c 90 
d3 02 00 00 
00 00 00 00 
d3 02 00 00
18 42 00 80 0e 00 32 30 31 35 30 34 32 31 30 30 30 32 45 00 
19 42 00 80 01 00 01 
20 42 00 80 08 00 44 57 52 2d 39 32 31 00 
26 42 00 80 03 00 43 31 00 
27 42 00 80 03 00 43 31 00 
21 42 00 80 02 00 02 1a 
22 42 00 80 04 00 c0 a8 7b fe 
10 42 00 80 0d 00 44 4c 4b 36 45 32 34 31 34 30 30 31 00 
01 42 00 80 02 00 24 6e 
00 f5 00 80 00 02 20 76
05 01 00 0c 43 76 20 e0 ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff 00 0c 43 76 20 77 00 0c 43 76
20 66 22 0c 00 00 00 01 61 01 55 77 a8 aa 8c 88
ff ff 0a 00 00 00 00 00 00 00 00 00 ff ff ff ff
06 06 07 07 08 08 09 09 09 08 08 08 07 07 07 07
07 07 07 06 06 06 06 07 07 07 08 08 80 ff ff ff
80 ff ff ff 00 00 ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff 20 ff
ff ff ff ff ff ff ff ff ff ff ff ff 08 08 08 08
04 00 07 07 04 00 08 08 04 00 07 07 04 00 ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
```

xabolcs commented:
(which continues as ...:)
```
37 42 00 80 12 00 47 31 53 4a 46 35 30 30 30 34 42 33 4a 30 31 00 55 53
30 42 00 80 06 00 e4 6f 13 26 d5 51
81 42 00 80 11 00 50 55 72 46 35 55 58 52 75 52 36 36 50 54 63 6b 00
83 42 00 80 11 00 76 71 47 65 46 36 45 33 38 32 39 6e 77 53 38 34 00
82 42 00 80 09 00 36 32 35 36 37 39 32 38 00
00 00 00 00
```

Which translates as
```
37 42 00 80 12 00: HEADER (4237 -> config data (id 37), 0012 -> size)
30 42 00 80 06 00: HEADER (4230 -> MAC (id 30), 0006 -> size)
81 42 00 80 11 00: HEADER (4281 -> config data (id 81), 0011 -> size)
83 42 00 80 11 00: HEADER (4283 -> config data (id 83), 0011 -> size)
82 42 00 80 09 00: HEADER (4282 -> config data (id 82), 0009 -> size)
00 00 00 00: END
```

Looks like DWR-921 doesn't have section id 62, which the DWR-116 has.
And DWR-921 have config sections which the DWR-116 doesn't have:
* id 37, length 0x12
* id 81, length 0x11, 0 terminated string?
* id 83, length 0x11, 0 terminated string?
* id 82, length 0x11, 0 terminated string?


@fid0did0, what does DWR-921's JBOOT Recovery command `info` say?

fid0did0 commented:
To compleate I push also the DWR-512 relevant config section:

```
0000e000: 02 02 24 2b e6 d4 ed 07 8f 02 00 00 ff ff 76 f3  ..$+..........v.
0000e010: 43 53 5f 3e 7f 02 00 00 00 00 00 00 7f 02 00 00  CS_>............
0000e020: 18 42 00 80 0e 00 32 30 31 35 30 31 31 36 30 30  .B....2015011600
0000e030: 30 33 45 00 19 42 00 80 01 00 01 20 42 00 80 0a  03E..B..... B...
0000e040: 00 44 57 52 2d 35 31 32 42 31 00 21 42 00 80 02  .DWR-512B1.!B...
0000e050: 00 fc ff 22 42 00 80 04 00 c0 a8 7b fe 10 42 00  ..."B......{..B.
0000e060: 80 0d 00 44 4c 4b 36 45 32 34 31 32 30 30 31 00  ...DLK6E2412001.
0000e070: 01 42 00 80 02 00 24 6e 30 42 00 80 06 00 78 54  .B....$n0B....xT
0000e080: 2e a0 78 6d 00 f5 00 80 00 02 50 53 00 01 00 50  ..xm......PS...P
0000e090: 18 21 d5 28 ff ff ff ff ff ff ff ff ff ff ff ff  .!.(............
0000e0a0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e0b0: ff ff 00 0c 43 33 52 77 00 0c 43 33 52 66 11 00  ....C3Rw..C3Rf..
0000e0c0: 24 10 00 01 40 01 55 77 a8 aa 8c 88 ff ff 0c 00  $...@.Uw........
0000e0d0: 00 00 00 00 00 00 00 00 ff ff ff ff 0b 0b 0c 0c  ................
0000e0e0: 0b 0b 0b 0b 0b 0b 0c 0c 0c 0c 10 10 11 11 11 12  ................
0000e0f0: 12 12 13 13 13 14 14 14 ff ff ff ff ff ff ff ff  ................
0000e100: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e110: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e120: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e130: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e140: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e150: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e160: ff ff ff ff ff ff ff ff 66 66 cc aa 88 66 cc aa  ........ff...f..
0000e170: 88 66 66 66 66 66 66 66 66 66 ff ff ff ff ff ff  .fffffffff......
0000e180: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e190: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1a0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1b0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1c0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1d0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1e0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e1f0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e200: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e210: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e220: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e230: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e240: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e250: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e260: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e270: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e280: ff ff ff ff ff ff ff ff ff ff 10 f5 00 80 0b 00  ................
0000e290: 45 31 35 43 45 33 43 32 35 39 00 00 00 00 00 ff  E15CE3C259......
0000e2a0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
0000e2b0: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  ................
```

xabolcs commented:
Awesome, it's the same pattern!
Looks like DWR-512 also introduces new section ids (at least: `F510`).

CHKDSK88 commented:
@mkresin 

Extractor ready.  Tested on DWR-116-A1 and A2.

xabolcs reviewed (commented):
Please mention in the commit message that the UART notation are different on the boards: it's `J1` on A1 and `JP1` on A2!

xabolcs commented on package/utils/jboot-tools/Makefile:
```diff
@@ -0,0 +1,28 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=jboot-tools
```
Reading only the package name I think the `jboot-tools` is related to the bootloader of the router.
Sure, the description clearly specifies that it will poke around the config partition.

How about putting `eeprom` in the name: `jboot-eeprom-tools`? Unsure about `fw`, `config` and others.

mkresin replied:
eeprom-tools doesn't really fit. It is similar to the uboot-envtools and fritz-tffs. But to be honest I don't really care about the name.

xabolcs replied:
>But to be honest, I don't really care about the name.

Ok, then keep `jboot-tools`' name as is. :)

xabolcs replied:
Ok, then keep this as is.

xabolcs commented on package/utils/jboot-tools/Makefile:
```diff
@@ -0,0 +1,28 @@
+  SECTION:=firmware
+  CATEGORY:=Firmware
+  DEPENDS:=@TARGET_ramips
+  TITLE:=Utilites for accessing JBOOT based D-link devices Calibration data
```
Same uppercase `L` in `D-Link`.

xabolcs commented on package/utils/jboot-tools/Makefile:
```diff
@@ -0,0 +1,28 @@
+
+define Package/jboot-tools/description
+ This package contains:
+ jboot_config_read.c: partially read the config partition of JBOOT based D-link devices.
```
Same uppercase `L` in `D-Link`.

xabolcs commented on package/utils/jboot-tools/src/jboot_config_read.c:
```diff
@@ -0,0 +1,427 @@
+/*
+ * jboot_config)read
```
Typo here: `)`

xabolcs reviewed (commented):
Suggestions. Typos.

CHKDSK88 commented:
Thanks for checking @xabolcs . I corrected D-Link names and my mistake in comment.

In package name case, I will wait for @mkresin opinion.

mkresin commented:
@fid0did0 @xabolcs 

Are you fine with the result?

xabolcs commented:
> Are you fine with the result?

Looks promising, but did a basic testing only. I have `A2 EU` revision.

I have two thoughts now:

The MAC addresses are swapped ...: WAN MAC = (WLAN MAC and LAN MAC) +1 instead of:
> 4. Set WLAN MAC the same as LAN MAC (WAN +1).

By the way, the magic at `02_network` and `10-rt2x00-eeprom` is unclear to me.
Will check the stock firmware, how it behaves on my `A2`.
It would nice to keep in sync the MACs for every interface with the stock firmware.


Another one is the model name I noted in the `dts`:
> What do you think about naming the router without the `A` revision prefix? Just as `D-Link DWR-116` and `dlink,dwr-116`?
> 
> I'm against the `A` because it's not a full revision string, as `A1`, `A2`, `A2EU` or `A2PL` are.
I'm OK with `D-Link DWR-116 A1/A2` and `dlink,dwr-116-a1-a2` too. :)

mkresin commented:
Use only `dlink,dwr-116-a1` as boardname/compatible string and "D-Link DWR-116 A1/A2" as model/DEVICE_TITLE.

We have similar examples in the tree where an image works fine with a later board revision. These board are using the first supported revision as boardname/compatible string as well.

xabolcs commented:
> Will check the stock firmware, how it behaves on my A2.

It works as like other revisions: WAN +1 = LAN = WLAN.


I'm entering in the second pass with the `02_network` and `10-rt2x00-eeprom`.

@mkresin, are there any way to define the offset (`0xE000` or `0x7FE000`) in the dts and reuse it in above scripts? Or label it as a named sub-partition of `config`?
Both scripts use that offset.
Or it is a allowed to use `/lib/firmware/$FIRMWARE` (`/lib/firmware/soc_wmac.eeprom` exactly) in `02_network`?
Will do some minor suggestion to the file additions in these scripts.


@CHKDSK88, another thing I found, that I don't see the failsafe phase of the boot through the status LED. Am I missing something?
When the boot is finished, the status LED is going to solid, but doesn't blinks before it:
- `. /etc/diag.sh; set_state upgrade` switches off my status LED instead of blinking
- `. /etc/diag.sh; set_state done` switches on my status LED
- defining a timer LED trigged, doesn't blinks the LED

xabolcs commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -425,6 +426,10 @@ ramips_setup_macs()
 		;;
+	dlink,dwr-116-a)
+		lan_mac=$(jboot_config_read -m -i /dev/mtd5 -o 0xE000)
+		wan_mac=$(macaddr_add "$lan_mac" 1)
```
`wan_mac` and `lan_mac` should be swapped and I suggest to use the `find_mtd_index` function.

Example based on board `wsr-1166` :
```
		local index="$(find_mtd_index "config")"
		wan_mac="$(jboot_config_read -m -i "/dev/mtd${index}" -o 0xE000)"
		lan_mac=$(macaddr_add "$wan_mac" -1)
```

mkresin commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -425,6 +426,10 @@ ramips_setup_macs()
 		wan_mac=$(mtd_get_mac_ascii factory wanmac)
 		;;
+	dlink,dwr-116-a)
+		lan_mac=$(jboot_config_read -m -i /dev/mtd5 -o 0xE000)
```
don't use a fixed mtd path. Something like the following should work:

```
lan_mac=$(jboot_config_read -m -i $(find_mtd_part "config") -o 0xE000)
```

I've seen the same in another file. Please fix it as well.

xabolcs commented on target/linux/ramips/base-files/etc/hotplug.d/firmware/10-rt2x00-eeprom:
```diff
@@ -41,6 +54,11 @@ board=$(board_name)
 "soc_wmac.eeprom")
 	case $board in
+	dlink,dwr-116-a)
+		wifi_mac=$(jboot_config_read -m -i /dev/mtd5 -o 0xE000)
```
For the `wifi_mac` I would implement another `jboot_xxxxxx_extract` function that extracts the *WAN* MAC from the "config" partition, and adds 1 to it (with the help of `. /lib/functions/system.sh`).

I would also save the `0xE000` offset to a local variable to avoid writing it twice. :)

xabolcs replied:
> ... I would implement another `jboot_xxxxxx_extract` function ...

In the previous version of `jboot-tools` there was a functionality like that? Wasn't it?

mkresin commented:
> @mkresin, are there any way to define the offset (0xE000 or 0x7FE000) in the dts and reuse it in above scripts? Or label it as a named sub-partition of config?
Both scripts use that offset.

Sure you can create a matching partition table. But that wouldn't match the stock firmware partition table. Hence I'm fine with using the offset at multiple places.

>Or it is a allowed to use /lib/firmware/$FIRMWARE (/lib/firmware/soc_wmac.eeprom exactly) in 02_network?

No, please don't do it. If the path ever change, it will be a nightmare to find such places.

> I don't see the failsafe phase of the boot through the status LED. Am I missing something?

Recently introduced kernel bug with LED blinking.

xabolcs commented:
> Sure you can create a matching partition table. But that wouldn't match the stock firmware partition table. Hence I'm fine with using the offset at multiple places.

I mean sub-partitioning, like the `jimage-fw` and `squasfs-split` in the `dmesg`:
```
[    0.587577] Creating 3 MTD partitions on "spi0.0":
[    0.597133] 0x000000000000-0x000000010000 : "jboot"
[    0.608350] 0x000000010000-0x0000007f0000 : "firmware"
[    0.620583] 2 jimage-fw partitions found on MTD device firmware
[    0.632457] 0x000000010000-0x00000015cdb4 : "kernel"
[    0.643939] 0x00000015cdb4-0x0000007f0000 : "rootfs"
[    0.655529] mtd: device 3 (rootfs) set to be root filesystem
[    0.667325] 1 squashfs-split partitions found on MTD device rootfs
[    0.679696] 0x00000044c000-0x0000007f0000 : "rootfs_data"
[    0.691923] 0x0000007f0000-0x000000800000 : "config"
```

I'm thinking of a `config` partition at `0x7f0000` with length `0x10000`, and a `config-psec` subpartition at `0x7fe000` with length ??? 
Is it possible?

The `jboot`, `firmware` and `config` partitions don't match the stock `Whole`, `Bootloader`, `Kernel`, `RootFS`, `UI` and `Config` partitions.
Only the the start positions match with the stock.

mkresin commented:
> I'm thinking of a `config` partition at `0x7f0000` with length `0x10000`, and a `config-psec` subpartition at `0x7fe000` with length ??? 
Is it possible?

IMHO it isn't worse the work.

> The `jboot`, `firmware` and `config` partitions don't match the stock `Whole`, `Bootloader`, `Kernel`, `RootFS`, `UI` and `Config` partitions.
Only the the start positions match with the stock.

I would expect the following partition mapping/matching:

Whole == complete flash, no idea why it exists
Bootloader == jboot
Kernel + RootFS + UI == firmware
config == config

CHKDSK88 commented:
@mkresin 

MAC swap (LAN, WAN) done. Now is the same like stock.
Boardname and model done.
Fixed mtd name.

xabolcs commented:
One step closer to the merge! :+1:

> Boardname and model done

"D-Link DWR-116 A1/2" is intended or just a typo? I'm missing an `A` before `2`.

CHKDSK88 commented:
@xabolcs 
Heh, typo. Morning distraction. ;)

Done.

mkresin commented:
Isn't it necessary to enable the ephy port 4 again since https://github.com/openwrt/openwrt/commit/d482356322c9910541a3118d049718b384a252bd?

CHKDSK88 commented:
@mkresin 
No.  I removed this section.

fid0did0 commented:
Hi all,
I currently tested on my dwr-921 the following patched files/functions:
 target/linux/ramips/base-files/lib/upgrade/platform.sh
target/linux/ramips/mt7620/config-4.14
target/linux/ramips/mt7620/config-4.9
tools/firmware-utils/src/mkdlinkfw-lib.c
tools/firmware-utils/src/mkdlinkfw-lib.h
tools/firmware-utils/src/mkdlinkfw.c

Currently I only have an issue after the fw installation. 
After mounting the rootfs I get:

```
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    5.973715] mount_root: no usable overlay filesystem found, using tmpfs overlay

....

[   15.548070] jffs2: jffs2_scan_eraseblock(): Magic bitmask 0x1985 not found at 0xyyyyyyyyy: 0x7d15 instead
[   15.613885] jffs2: jffs2_scan_eraseblock(): Magic bitmask 0x1985 not found at 0xyyyyyyyyy: 0xcc2e instead
[   15.673856] jffs2: jffs2_scan_eraseblock(): Magic bitmask 0x1985 not found at 0xyyyyyyyyy: 0xe3c5 instead
```
I checked the fw image and I can see the 0xdeadc0de @ 
0x37F000  DE AD C0 DE
0x380000	  DE AD C0 DE

I need to check if the 0xdeadc0de is properly flashed into the device.
Bye.

CHKDSK88 commented:
@fid0did0

Did You flash over stock web gui? 
JBOOT rescue flash make it wrong.

fid0did0 commented:
I'm using the JBOOT rescue flash.
Why it should be make it wrong?
I'm expecting that it flashes the amount of data reported by the header.
In this case, why not extend the IMAGE_ROOTFS with the sector including the 0xdeadc0de?

Something like that:
```
+define Build/mkdlinkfw
+	-$(STAGING_DIR_HOST)/bin/mkdlinkfw \
+		-m $(DLINK_ROM_ID) -f $(DLINK_FAMILY_MEMBER) \
+		-k $(IMAGE_KERNEL) \
+		-r $(( $(IMAGE_ROOTFS)+$(DEADCODE_SIZE) )) \
+		-o $@ \
+		-j \
+		-s $(DLINK_FIRMWARE_SIZE) \
+		$(if $(findstring sysupgrade,$(word 1,$(1))),-S)
+endef
```

CHKDSK88 commented:
> I'm using the JBOOT rescue flash.
> Why it should be make it wrong?

Because JBOOT have problem with big partitions.

>I'm expecting that it flashes the amount of data reported by the header.
>In this case, why not extend the IMAGE_ROOTFS with the sector including the 0xdeadc0de?

DEADC0DE is included. Whole partition isn't erased. That's JBOOT issue.

mkresin commented:
> DEADC0DE is included. Whole partition isn't erased. That's JBOOT issue.

0xDEADC0DE is exactly used for the "Whole partition isn't erased" issue.

During firstboot the squashfs header is evaluated. The beginning of the next block after $SIZE_FROM_SQUASHFS_HEADER is check for the 0xDEADC0DE marker. If the marker is found, the remaining flash space of the (firmware) partition beginning and including the block with the  0xDEADC0DE is formated as jffs2 rootfs_data.

Bootloader validating the rootfs checksum have issues with the 0xDEADC0DE marker, since it is gone after the first boot => checksum doesn't match any longer. One workaround used is to set the rootfs size to a value not covering the 0xDEADC0DE marker.

Depending on which/how headers are evaluated by the firmware upgrade routine even this can fail. Some of the routines only write as much bytes as defined in the related header.

I've no idea how it is done here and I'm to busy (maybe to lazy) to read the code. There are quite some more known workarounds to bypass as the limitation. If this is identified as issues, we can think about which one would fit here.

I added this commit to [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary), albeit I'm not fully satisfied. I never liked the mktplinkfw approach to add the padding + 0xDEADC0DE marker within the tool instead of using the pad-rootfs image build step. a) it is duplicate code, b) it is hard to get what is really included in the (crypted) image.

But I really want to bring this PR to an end and therefore do not request this change, even if it would be nice to see it.

Please close the PR if you don't intend to touch the code once more. Maybe my explanation about image size headers and the evaluation required another code change.

fid0did0 commented:
Hi mkresin and thanks for the clarification.

In my opinion it should be not so complex to appla a strategy to extend the fw image with the 0xDEADC0DE blocks and after the generation of the jffs2 change the checksum like others routers do:
openwrt/target/linux/ramips/base-files/lib/preinit/04_handle_checksumming

If nobady will do it before, I will work on this in some months.

Bye.

CHKDSK88 commented:
@mkresin 
 I made one more change in target/linux/ramips/image/mt7620.mk.

I check if is possible to make images with using the pad-rootfs image build step and let You know.

CHKDSK88 commented:
@mkresin 

Done.

@fid0did0 

Fixed factory. No You can upload it thru jboot rescue.

fid0did0 commented:
I tested the pad-rootfs factory image.
Now the jffs2 should be fine:

```
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    7.669874] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[    7.712673] urandom-seed: Seed file not found (/etc/urandom.seed)
[ ... ]
[   16.332938] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   16.383381] jffs2_build_filesystem(): unlocking the mtd device... [   16.422380] done.
[   16.426218] jffs2_build_filesystem(): erasing all blocks after the end marker...
[ ... ]
[  321.369176] jffs2: notice: (898) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
```

mkresin commented:
@CHKDSK88 I should have your latest version in my staging tree. Would you please build an image from my staging tree and test this one.

I'm still curious that you don't need to enable port4. It looks like your OpenWrt  isn't up to date and misses the commit that makes enabling port4 mandatory. With the image build from my tree, we will see if it is required or not.

CHKDSK88 commented:
@mkresin 

>I should have your latest version in my staging tree. Would you please build an image from my staging tree and test this one.

I build directly from Your staging tree and upload to my DWR-116.

>I'm still curious that you don't need to enable port4. It looks like your OpenWrt isn't up to date and misses the commit that makes enabling port4 mandatory. With the image build from my tree, we will see if it is required or not.

Port 4 still works perfectly.
Maybe dtb entry concerns rgmii interface only?

What You think? Is now the moment to close this PR?

mkresin commented:
> What You think? Is now the moment to close this PR?

Yes it is.

CHKDSK88 commented:
@mkresin 
Thank You for Your advices and helping to make it properly.

@fid0did0 @xabolcs 
Thank for Your support. Your participation was very important durring all work.

Now the gate for the rest JBOOT devices are open.

fid0did0 commented:
> I'm still curious that you don't need to enable port4. It looks like your OpenWrt isn't up to date and misses the commit that makes enabling port4 mandatory. With the image build from my tree, we will see if it is required or not.

After a fast check in the driver I see that the port 4 is enabled if it is configured as ephy independently from the status in dts. As done by CHKDSK88 in 39b5218.

I need this enable because the port is disabled when the OS start.

The point is that it seems that the driver doesn't support the status on the dts.

In my opinion as it is now it is a minor issue and we can keep it.

Let me know what you think about.
Bye.

mkresin commented:
> After a fast check in the driver I see that the port 4 is enabled if it is configured as ephy independently from the status in dts

Yeah, that was the point I missed and it is similar to the mdio node as I head to learn with [FS#1428](https://bugs.openwrt.org/index.php?do=details&task_id=1428). It also means that my fix of the "invalid port id 4" error doesn't work at all.

> The point is that it seems that the driver doesn't support the status on the dts. In my opinion as it is now it is a minor issue and we can keep it.

Agreed. I'm not going to touch it without having hardware to test it. I'm not keen to break it again.
